README
PDL::Drawing::Prima
===================

A PDL interface to the Prima drawing commands.

Buried deep inside the wonderful, Perl-centric Prima toolkit is a set of
cross-platform C functions that handle all of the drawing operations. PDL's
strength is its threading engine, and this module provides a fully threaded
subset of Prima drawing functions. As such, this module provides a way to
quickly perform many drawing operations on Prima Widgets.

At the moment, the bindings are only for screen-drawn objects. I started
work on the Postscript drawing interface but did not get very far. That may
get more attention in the future, likely to satisfy the file-output demands
of PDL::Graphics::Prima.

For more information, see the documentation for PDL::Drawing::Prima.


INSTALLATION
============

There are a number of ways to install this module. The simplest is to
install it using CPAN:

	cpan PDL::Drawing::Prima

If you have your hands on the actual source code (or you "look"ed at this
module from the CPAN shell), you would do the following:

	perl Makefile.PL
	make
	make test
	make install

There are almost no tests for this distribution at the moment. I would like
to build a number of tests by emulating the PDL behavior with Perl loops
and comparing the rasterized output, but I have not gotten that far yet.
If this project gets a bit of momentum behind it, I truly hope that the
test suite gets some much-needed attention.


DEPENDENCIES
============

The installation of this module depends on a working C compiler. It also
depends on the Perl Data Language and the Prima GUI toolkit. You can
install both of these using CPAN. If you manage to get those installed
without trouble, installng this module should work fine unless you are
working with the stock Debian package of Prima. That package was improperly
built and does not list the correct location of Prima's CORE directory. That
makes it impossible to link against apricot.h (Prima's main library). I may
be able to write a special exception into the Makefile.PL for this, but I
haven't gotten around to it yet.


CODE REPOSITORY
===============

This project is being actively developed at Github. Find the latest at:

 https://github.com/run4flat/PDL-Drawing-Prima


COPYRIGHT AND LICENCE
=====================

Copyright (C) 2011, David Mertens

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

Author: David Mertens <dcmertens.perl.csharp@gmail.com>

Note that my email is intentionally obfuscated. Knowing that I am a
Perl programmer, you can probably remove the part that does not belong.


Changes
- fixes to build for distro

0.17 Feb 27, 2023

   Convert to EUMM

0.16 Jan 16, 2023

   Added pictures to the docs

0.15 Dec 24, 2022

  * Dmitry Karasik takes over the project.
    - Compile with latest Prima
    - Update links
    - Switch from Travis CI to Github Actions
    - Run tests without X11

0.14_01 MTH XX, 2022

0.14 Feb 13, 2022

  [ENHANCEMENTS]
  
  * Now includes a real test! t/10 uses the Prima/Perl interface and the
    pdl interface to draw what should be identical things---lines of
    various colors. It then performs a pixel-by-pixel comparison to make
    sure you get the exact same result. Much more needs to be written
    but now at least I have the template.

  [DEPENDENCIES]
  
  * Bumped required verson of PDL to v2.037. Should have done this with the
    release of v0.13, but that was a bit hasty. Oops. This is the required
    minimum version mostly because some versions of PDL earlier than
    v2.037 made changes to PDL::PP that broke the compilation of this
    distribution. The changes to PDL::PP were fixed with v2.037.
  
  [BUG FIXES]
  
  * Fixed a minor bug that caused PDL::Drawing::Prima::Utils to not compile
    on some platforms. Should have used pp_line_numbers, but rolled my own
    in a few spots for some reason.

0.13 Feb 09, 2022

  [BUG FIXES]
  
  * Prima changed fillWinding to fillMode, which required a change to
    PDL::Drawing::Prima's interface. Thanks to dk for this fix!

  [DEPENDENCIES]
  
  * Bumped required verson of Prima to 1.57

0.12 Aug 01, 2016

  [BUG FIXES]
  
  * PDL::PP now requires a call to pp_done, which was missing in Utils

0.11 Nov 07, 2015

  [BUG FIXES]
  
  * Fixed backwards-compatible support code, which wasn't actually fully
    backwards compatible as written.
  
  * Update Build.PL to work with newest PDL.
  
  [DEVELOPMENT]
  
  * Added Travis support for github repository


0.10 Oct 28, 2013

  [ENHANCEMENTS]
  
  * Added backwards-compatible support for PDL v2.007.
  
  * Hacked an improved symbol drawing for crosses and triangles. It sometimes
    happened that the bottom leg of a cross symbol wasn't straight down due to
    rounding a floating-point number that is slightly less than the integer it
    was supposed to be. This hack makes it work slightly better, though an even
    better algorithm would be preferred.
  
  * Added a Changes file.
  
  * Revised PDL::Drawing::Prima::Utils::minmaxforpair() to use standard
    functions in math.h for determining whether a value is finite or not.
  
  [BUG FIXES]
  
  * Fixed corner-case in PDL::Drawing::Prima::Utils::hsv_to_rgb.



0.09 Jan 4, 2013

  [BUG FIXES]
  
  * Fixed Build.PL's linking; now should work on Cygin and Strawberry.



0.08 Sep 9, 2012

  [BUG FIXES]
  
  * Fixed Windows runtime link error.
  
  [DEPENDENCIES]
  
  * Bumped required verson of Prima to 1.35



0.07 Aug 6, 2012
  
  [API CHANGES]
  
  * Deprecated PDL::Drawing::Prima::pdl_of_handles_for()
  
  [DIFFERENT BEHAVIOR]
    
  * Drawing operations are now atomic!! Now, when you call pdl-based drawing
   function on a Prima drawable, the drawable's state will be restored to its
   original state before the function returns.
  
  [ENHANCEMENTS]
  
  * Documentation for the properties supported by a drawing operation are now
    guaranteed to be correct because they are extracted directly from the
    documentation.
  
  [DEPENDENCIES]
  
  * Added minimum version for Module::Build (0.28)
  
  * Bumped minimum version for PDL to 2.4.10
  
  * Added runtime Prima version check

  [NEW FEATURES]
  
  * Added fillPattern property to applicable drawing functions.



0.06 Jul 12, 2012

  XXX WORKING HERE



0.05 Jul 6, 2012

  XXX WORKING HERE



0.04 Jul 5, 2012

  XXX WORKING HERE



0.03 Jun 25, 2012

  XXX WORKING HERE



0.02 Mar 15, 2012

  XXX WORKING HERE



0.01 Feb 26, 2012

  XXX WORKING HERE


Build.PL
use Module::Build;
use strict;
use warnings;
use PDL::Core::Dev;
use Prima::Config;
use Config;

# These are lifted from PDL::Core::Dev, but use a less redundant style:
my $inc = PDL_INCLUDE;
$inc =~ s/-I//g;
$inc =~ s/"//g;
my @incs = ($inc, @{$Prima::Config::Config{incpaths}});

# Explicitly set the version here; all .pm.PL files will use this:
my $version = '0.16';

my $build = Module::Build->new(
	module_name => 'PDL::Drawing::Prima',
	dist_version_from => 'lib/PDL/Drawing/Prima.pm.PL',
	dist_version => $version,
	license  => 'perl',
	configure_requires => {
		PDL   => '2.037',
		Prima => '1.57',
		# meta_merge was added in v 0.28:
		'Module::Build'		=> 0.28,
	},
	build_requires => {
		# working here - I need to check the smoke reports to hone in on a
		# minimum required version:
		'ExtUtils::ParseXS'	=> '3.0',
	},
	requires => {
		# Due to the use of Prima, I need at least 5.12:
		perl  => '5.12.0',
		PDL   => '2.037',
		Prima => '1.57',
	},
	needs_compiler => 1,
	include_dirs => \@incs,
	meta_merge => {
		resources => {
			repository => 'http://github.com/dk/PDL-Drawing-Prima',
			bugtracker => 'http://github.com/dk/PDL-Drawing-Prima/issues',
		},
		provides => {
			'PDL::Drawing::Prima' => {
				file => 'lib/PDL/Drawing/Prima.pm.PL',
				version => $version,
			},
			'PDL::Drawing::Prima::Utils' => {
				file => 'lib/PDL/Drawing/Prima/Utils.pm.PL',
				version => $version,
			},
		},
		no_index => {
			file => ['README.pod'],
			directory => [qw(examples t)],
		},
	},
);

# Link to Prima's dynamic library if there are any linker settings.
# (Big shout out to Dmitry, Rob, and Chris for helping me solve this!!)
$build->extra_linker_flags(split(/\s+/, $Prima::Config::Config{libs}))
	if $Prima::Config::Config{libs};

$build->add_build_element('png');
$build->create_build_script;
